### ISSUE TITLE
JVM VerifyError after transformation from .class via shimple

### ISSUE THREAD

**Issue by Waldo1637** [Sable/soot#385]
When running soot via Shimple with compiled class files as source, some try-catch blocks are not processed correctly. Specifically, when a variable is assigned inside the try block (before a statement that the ThrowAnalysis directs to the catch block) and used after the try-catch, the output class file fails the JVM bytecode verification with an error such as:

Exception in thread "main" java.lang.VerifyError: (class: VerifyErrorTest, method: readFromConsole signature: ()Ljava/lang/String;) Accessing value from uninitialized register 1
        at java.lang.Class.getDeclaredMethods0(Native Method)
        at java.lang.Class.privateGetDeclaredMethods(Unknown Source)
        at java.lang.Class.getMethod0(Unknown Source)
        at java.lang.Class.getMethod(Unknown Source)
        at sun.launcher.LauncherHelper.getMainMethod(Unknown Source)
        at sun.launcher.LauncherHelper.checkAndLoadMain(Unknown Source)

However, this does not seem to happen when soot is given the java source as input.

A simple test case is available here: https://www.dropbox.com/sh/gvur3vfkykkqd7i/AABAkdu4AUtyku6IVg3qpmOTa?dl=0

You will see the error when trying to run the output given by the following:

```
C:\Soot_VerifyError> java soot.Main -via-shimple -pp -cp class VerifyErrorTest
```

However, the output given by the following passes JVM verification:

```
C:\Soot_VerifyError> java soot.Main -via-shimple -pp -cp src VerifyErrorTest
```

In that link you will also find an image of the ExceptionalBlockGraph and the jimple that are given when VerifyErrorTest.class is used as input to soot. On examining the jimple file, you can see the statement on line 34 is what causes the VerifyError because r0_2 may not be initialized due to the exception handler covering label1 to label2.

**Comment by Waldo1637** [comment]
Upon further tests, I have found that this bug is not dependent upon whether java source or compiled bytecode is used. I have added another test case VerifyErrorTest2.java which causes the VerifyError when executed regardless of whether the source or the bytecode is used as input to Soot.

**Comment by ericbodden** [comment]
@Waldo1637 can you post the Jimple code of the method in question? (readFromConsole)

You can use -f J to produce the jimple code.

**Comment by Waldo1637** [comment]
Here is the jimple created using the **class file** as input:

```java
private static java.lang.String readFromConsole()
{
	java.lang.String r0, r0_1, r0_2, r0_3;
	java.util.Scanner $r2;
	java.io.InputStream $r3;
	java.lang.Exception $r4;

	r0 = null;

 label1:
	$r2 = new java.util.Scanner;

	$r3 = <java.lang.System: java.io.InputStream in>;

	r0_2 = r0;

	specialinvoke $r2.<java.util.Scanner: void <init>(java.io.InputStream)>($r3);

	r0_1 = virtualinvoke $r2.<java.util.Scanner: java.lang.String nextLine()>();

 label2
	r0_3 = r0_1;

	goto label4;

 label3
	$r4 := @caughtexception;

	r0_3 = r0_2;

 label4:
	return r0_3;

	catch java.lang.Exception from label1 to label2 with label3;
}
```

Here is the jimple created using the **source code** as input:

```java
private static final java.lang.String readFromConsole()
{
	java.lang.String s, s_2;
	java.util.Scanner temp$0;
	java.io.InputStream temp$1;
	java.lang.Exception e

	s = null;

 label1:
	temp$0 = new java.util.Scanner;

	temp$1 = <java.lang.System: java.io.InputStream in>;

	specialinvoke temp$0.<java.util.Scanner: void <init>(java.io.InputStream)>(temp$1)

	s_2 = virtualinvoke temp$0.<java.util.Scanner: java.lang.String nextLine()>()

	goto label3

 label2:
	e := @caughtexception

	s_2 = s

 label3:
	return s_2

	catch java.lang.Exception from label1 to label2 with label2;
}
```

The full Jimple output can be found via the Dropbox link I gave earlier. In that folder you can also find the Jimple output for the VerifyErrorTest2 class for which Soot created unverifiable code regardless of whether source or bytecode was used as input.

**Comment by Waldo1637** [comment]
I think the solution to this problem may be to change the ExceptionalUnitGraph to insert a new edge to the handler from the unit preceding the trap. This would allow phi node removal to propagate the assignment statement to where it should be: before the trap.

For example, here is the original ExceptionalBlockGraph for readFromConsole():
![exceptionalblockgraph](https://cloud.githubusercontent.com/assets/4001421/7287373/993de6fe-e919-11e4-966d-300fbedf664c.png)

And here is what the graph should look like with the new edge:
![exceptionalblockgraph_desired](https://cloud.githubusercontent.com/assets/4001421/7287493/c11db16c-e91a-11e4-943c-b714a1c3fffc.png)


This could be as simple as changing the condition on line 355 of ExceptionalUnitGraph.java from
```java
!catchableAs.getCaught().equals(ThrowableSet.Manager.v().EMPTY)
```
to
```java
unit.equals(trap.getBeginUnit()) || !catchableAs.getCaught().equals(ThrowableSet.Manager.v().EMPTY)
```
However, I am quite unfamiliar with that class so I'm not sure if that change would change the graph correctly without breaking anything else.

**Comment by ericbodden** [comment]
Thanks for looking into this so thoroughly.

I am not 100% sure, though, that this change will be correct in all the possible cases.  Also I don't quite understand how the added control-flow graph edge is responsible for fixing the bytecode-verification problem you saw. How are the two connected? Do you know?

**Comment by Waldo1637** [comment]
When Soot removes the phi nodes, it propagates new assignment statements backwards along the graph edges to the blocks where the original assignment statements are located (the ones that contributed to the phi node) and inserts the new assignment statement at the end of that basic block. In the first Jimple code example I posted on Apr 10, this is where the line `r0_2 = r0;` comes from. Due to how the basic blocks are formed, this line is inserted within the trap even though the original assignment statement that it belongs with is located before the trap. The JVM verifier then asserts that `r0_2` may not have been initialized before its use (under label3) because any statement in the trap before `r0_2 = r0;` may throw an error. However, when we construct the graph as I suggested, the assignment statement is able to propagate outside of the trap. Thus, the verifier can assert that `r0_2` has been initialized before its use.

I think this suggested graph formation aligns more closely to the assumptions the JVM verifier makes about control flow. It checks the path in which none of the statements in the trap execute. That is, control flows directly from the statement immediately prior to the trap into the exception handler. Thus, when Soot forms the CFG, shouldn't it make the same assumption?
